optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java

```diff
@@ -0,0 +1,158 @@
+import io.micrometer.core.instrument.search.Search;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+
+public class StatisticRegistry extends SimpleMeterRegistry implements PhaseLifecycleListener, SolverEventListener {
+
+    List<Consumer<Long>> stepMeterListenerList = new ArrayList<>();
+    List<Consumer<Long>> bestSolutionMeterListenerList = new ArrayList<>();
+    long bestSolutionChangedTimestamp = Long.MIN_VALUE;
+    ScoreDefinition scoreDefinition;
+
+    public StatisticRegistry(DefaultSolver solver) {
+        scoreDefinition = solver.getSolverScope().getScoreDefinition();
+    }
+
+    public void addListener(SolverMetric metric, Consumer<Long> listener) {
+        if (isMetricBestSolutionBased(metric)) {
+            bestSolutionMeterListenerList.add(listener);
+        } else {
+            stepMeterListenerList.add(listener);
+        }
+    }
+
+    private boolean isMetricBestSolutionBased(SolverMetric metric) {
+        switch (metric) {
+            case BEST_SCORE:
+            case BEST_SOLUTION_MUTATION:
+            case CONSTRAINT_MATCH_TOTAL_BEST_SCORE:
+            case PICKED_MOVE_TYPE_BEST_SCORE_DIFF:
+                return true;
+            case MEMORY_USE:
+            case STEP_SCORE:
+            case ERROR_COUNT:
+            case SOLVE_LENGTH:
+            case MOVE_COUNT_PER_STEP:
+            case SCORE_CALCULATION_SPEED:
+            case PICKED_MOVE_TYPE_STEP_SCORE_DIFF:
+            case CONSTRAINT_MATCH_TOTAL_STEP_SCORE:
+                return false;
+            default:
+                throw new IllegalStateException("SolverMetric (" + metric + ") does not have a case");
+        }
+    }
+
+    public Set<Meter.Id> getMeterIds(SolverMetric metric, Tags runId) {
+        return Search.in(this).name(name -> name.startsWith(metric.getMeterId())).tags(runId)
+                .meters().stream().map(meter -> meter.getId())
+                .collect(Collectors.toSet());
+    }
+
+    public void extractScoreFromMeters(SolverMetric metric, Tags runId, Consumer<Score> scoreConsumer) {
+        String[] labelNames = scoreDefinition.getLevelLabels();
+        Number[] levelNumbers = new Number[labelNames.length];
+        for (int i = 0; i < labelNames.length; i++) {
+            Gauge scoreLevelGauge = this.find(metric.getMeterId() + "." + labelNames[i]).tags(runId).gauge();
+            if (scoreLevelGauge != null) {
+                levelNumbers[i] = scoreLevelGauge.value();
+            } else {
+                return;
+            }
+        }
+        scoreConsumer.accept(scoreDefinition.fromLevelNumbers(0, levelNumbers));
+    }
+
+    public void getGaugeValue(SolverMetric metric, Tags runId, Consumer<Number> gaugeConsumer) {
+        getGaugeValue(metric.getMeterId(), runId, gaugeConsumer);
+    }
+
+    public void getGaugeValue(String meterId, Tags runId, Consumer<Number> gaugeConsumer) {
+        Gauge gauge = this.find(meterId).tags(runId).gauge();
```

Comment:
Dtto. We do not need to look them up all the time, we can gradually build a cache.
(I'm of course assuming that the references do not change.)